### PR TITLE
fix(DomSpectator): setInput not trigger angular hooks

### PR DESCRIPTION
**Pull request by ntvsx193**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/spectator/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
The `setInput` not call angular hooks on testing component.

Issue Number: #461 

## What is the new behavior?
The `setInput` call angular hooks on testing component now.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

The `setInput` call root `detectChanges` from `fixture` now.

### CODE REVIEW

File: projects/spectator/jest/test/form-input/form-input.component.spec.ts [hunk 1]
```diff
@@ -4,53 +4,51 @@ import { createHostFactory, SpectatorHost, Spectator, createComponentFactory } f
 import { FormInputComponent } from '../../../test/form-input/form-input.component';
 
 describe('FormInputComponent', () => {
-  let host: SpectatorHost<FormInputComponent>;
-  const createHost = createHostFactory({
```
Comment (NetanelBasal) [inline projects/spectator/jest/test/form-input/form-input.component.spec.ts:8]: I'm not sure what these changes are.

File: projects/spectator/test/form-input/form-input.component.spec.ts [hunk 2]
```diff
@@ -4,50 +4,51 @@ import { createHostFactory, SpectatorHost, Spectator, createComponentFactory } f
 import { FormInputComponent } from './form-input.component';
 
 describe('FormInputComponent', () => {
-  let host: SpectatorHost<FormInputComponent>;
```
Comment (NetanelBasal) [inline projects/spectator/test/form-input/form-input.component.spec.ts:7]: I'm not sure what these changes are.

File: projects/spectator/test/set-input/set-input.component.spec.ts [hunk 3]
```diff
@@ -6,52 +6,83 @@ import { SetInputComponent } from './set-input.component';
...
-  });
+    beforeEach(() => {
+      spectator = createComponent();
+    });
 
-  it('should be update input from object', () => {
-    spectator.setInput({ one: '1' });
+    it('should be set value from OnInit angular hook', () => {
+      expect(spectator.component.fromInit).toBe('initValue');
+    });
```
Comment (NetanelBasal) [inline projects/spectator/test/set-input/set-input.component.spec.ts:22]: I'm not sure what these changes are.

File: projects/spectator/test/form-input/form-input.component.spec.ts (hunk 2 above)
Comment (ntvsx193) [reply projects/spectator/test/form-input/form-input.component.spec.ts:7]: 1. On my opinion you have strange example. You have `enableSubnet` input as static `true`. After that you change `enableSubnet` through `setInput`. There you change input of internal component `<app-form-input>` and not host component. Thats why I changed this test. I used `createHost` and `hostProps`.
 2. You had 2 describes with same name. I refactored the code to one top-level describe and 2 more different describes inside it.

File: projects/spectator/test/set-input/set-input.component.spec.ts (hunk 3 above)
Comment (ntvsx193) [reply projects/spectator/test/set-input/set-input.component.spec.ts:22]: I want to test 2 cases. I added 2 describes:
- defaults case (detectChanges = true)
- detectChanges = false

File: projects/spectator/jest/test/form-input/form-input.component.spec.ts (hunk 1 above)
Comment (ntvsx193) [reply projects/spectator/jest/test/form-input/form-input.component.spec.ts:8]: https://github.com/ngneat/spectator/pull/465#discussion_r684246251

File: projects/spectator/test/form-input/form-input.component.spec.ts (hunk 2 above)
Comment (ntvsx193) [reply projects/spectator/test/form-input/form-input.component.spec.ts:7]: If you see how to see is better, could you please share your thoughts?

**Comment by ntvsx193** [comment]
What about this pull request?

**Comment by NetanelBasal** [comment]
I will revisit it soon.

File: projects/spectator/src/lib/base/dom-spectator.ts [hunk 4]
```diff
@@ -97,7 +97,8 @@ export abstract class DomSpectator<I> extends BaseSpectator {
   public setInput<K extends keyof I>(input: K, inputValue: I[K]): void;
   public setInput(input: any, value?: any): void {
     setProps(this.instance, input, value, false);
-    this.debugElement.injector.get(ChangeDetectorRef).detectChanges();
+    setMarkForCheck(this.debugElement);
```
Comment (NetanelBasal) [inline projects/spectator/src/lib/base/dom-spectator.ts:100]: Invoking markForCheck when setting an input doesn't make sense to me. Why should we mark the whole tree when changing an input?

Comment (ntvsx193) [reply projects/spectator/src/lib/base/dom-spectator.ts:100]: We mark node for check need `detectChanges` and we run root `detectChanges` from `fixture`.
In this version, the all default hooks starts automatic in three if it need.
In previous version, the caller `detectChanges` running DC only on focused node.

**Comment by ntvsx193** [comment]
Will you merge these changes?
This is a very important fix for my team. We have 25,000 tests

**Comment by ntvsx193** [comment]
Fixed via #539

**Pull request closed without merging**